### DISCUSSION TITLE
Why Bun Give Unsorted Output Files in Read Directory?

### DISCUSSION THREAD

**Discussion by kresnasatya** [oven-sh/bun#14383] [Q&A] [answered]
Hi everyone! I create a script to get list of files inside of directory named `books`. Here's the directory structure:

```
- books
   - a.md
   - b.md
   - c.md
   - d.md
   - z.md
```

Here's the script.

```js
// index.js
import path from 'node:path';
import fs from 'node:fs/promises';

(async () => {
    const files = await fs.readdir(path.join(process.cwd(), 'books'));
    for (const file of files) {
        console.log(file);
    }
})();
```

If I run `node index.js` I get the sorted output files (ascending).

```bash
// node index.js
a.md
b.md
c.md
d.md
z.md
```

Meanwhile, if I run `bun index.js` I get unsorted output files.

```bash
// bun index.js <= Run this command many times and I get the exact output below.

b.md
z.md
c.md
d.md
a.md
```

I'm using Bun version 1.1.26 on macOS M1. I'm curious why does it happened? Is that design by default or is it bug?

**Comment by Jarred-Sumner** [comment, marked as answer]
This is consistent with Node.js - the operating system determines what sorting order to use, which is essentially random

**Reply by kresnasatya** [reply]
Thanks for the answer, Jarred.

**Reply by alexeyten** [reply]
NodeJS (well, actually libuv) *does* sort files on Linux (Mac), but not on Windows.

See:
https://github.com/nodejs/node/issues/3232
https://github.com/libuv/libuv/blob/v1.x/src/unix/fs.c#L580

I'm not saying that Bun should. I'm personally against implicit sorting.

**Reply by kresnasatya** [reply]
Thanks for detail information, alex.
